pkg/lib/synthesize/generate.go

```diff
@@ -0,0 +1,174 @@
+package synthesize
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/yaml"
+
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+)
+
+var CsvInferenceFileNotFoundError = errors.New("a file that contains non-inferable metadata could not be found")
+
+const (
+	SynthesizedCSVFileName string = "olm.clusterserviceversion.yaml"
+	DefaultPermission             = 0644
+)
+
+// GenerateCSV generates a CSV for the bundle based on the resource within the bundle directory.
+func GenerateCSV(bundleDir string) error {
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	// find annotations.yaml file in the bundle dir.
+	var annotations bundle.AnnotationMetadata
+	var annotationsFilePath string
+	if err := filepath.Walk(bundleDir, func(path string, info os.FileInfo, err error) error {
+		if annotations.IsComplete() {
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		annotationsFilePath = path
+		content, err := ioutil.ReadFile(annotationsFilePath)
+		if err != nil {
+			return err
+		}
+
+		// tries to parse the content as annotations else go to the next file.
+		annotations, _ = bundle.GetAnnotations(content)
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if !annotations.IsComplete() {
+		return fmt.Errorf("failed to find annotations YAML file in the bundle")
+	}
+
+	s, err := SynthesizeCsvFromDirectories(filepath.Join(bundleDir, annotations.GetManifestsDirName()),
+		filepath.Join(bundleDir, annotations.GetMetadataDirName()))
+	if err != nil {
+		return err
+	}
+
+	csv, err := s.GenerateCSV()
+	if err != nil {
+		return err
+	}
+
+	// generate CSV file to the outputDir.
+	manifestsDir := path.Join(bundleDir, annotations.GetManifestsDirName())
+	if _, err := os.Stat(manifestsDir); os.IsNotExist(err) {
```

Comment:
No need to do this, os.Mkdirall is a no-op if the file already exists and is a directory.